core/storage/user/gae_models.py

```diff
@@ -97,6 +97,61 @@ def get_by_normalized_username(cls, normalized_username):
             cls.normalized_username == normalized_username).get()
 
 
+class ActivitiesCompletedByLearnerModel(base_models.BaseModel):
+    """Keeps track of all the explorations and collections completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    # IDs of all the explorations completed by the user.
+    completed_exp_ids = ndb.StringProperty(repeated=True, indexed=True)
+    # IDs of all the collections completed by the user.
+    completed_collection_ids = ndb.StringProperty(repeated=True, indexed=True)
+
+
+class IncompleteCollectionsModel(base_models.BaseModel):
+    """Keeps track of all the collections currently being completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    incomplete_collection_ids = ndb.StringProperty(
+        repeated=True, indexed=True)
+
+
+class IncompleteExplorationsModel(base_models.BaseModel):
+    """Keeps track of all the explorations currently being completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    # A list which stores the explorations partially completed by the learner.
+    # Each item in the list is a json object keyed by exploration id and the
+    # value corresponds to a few extra details -
+    # timestamp: When did the learner leave the exploration.
+    # version: Version of the exploration.
```

Comment:
Clarified the definition. Yes, currently it saves the last played status of the exploration.Not sure whether this should happen though. Saving an older information can lead to potential problems such a saving a state that doesn't exist in the new version. So I think it would be best to save only if the version is higher, right?